File: xeg-upgrader/python/src/xeg_upgrader/evagene_client.py
"""Thin client for the subset of the Evagene REST API that the upgrader uses.

One method per endpoint; no orchestration (that lives in ``app``).  The
app layer talks to the :class:`EvageneApi` protocol so tests supply a
fake without a live HTTP layer.
"""

from __future__ import annotations

from typing import Any, Protocol

from .http_gateway import HttpGateway, HttpMethod

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class EvageneApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class EvageneApi(Protocol):
    def create_pedigree(self, display_name: str) -> str: ...

    def import_xeg_parse_only(self, pedigree_id: str, xeg_xml: str) -> dict[str, Any]: ...

    def import_xeg(self, pedigree_id: str, xeg_xml: str) -> None: ...

    def delete_pedigree(self, pedigree_id: str) -> None: ...


class EvageneClient:
    def __init__(self, *, base_url: str, api_key: str, http: HttpGateway) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def create_pedigree(self, display_name: str) -> str:
        payload = self._request_json("POST", "/api/pedigrees", {"display_name": display_name})
        return _require_str(payload, "id")

    def import_xeg_parse_only(self, pedigree_id: str, xeg_xml: str) -> dict[str, Any]:
        path = f"/api/pedigrees/{pedigree_id}/import/xeg?mode=parse"
        return self._request_json("POST", path, {"content": xeg_xml})

    def import_xeg(self, pedigree_id: str, xeg_xml: str) -> None:
        path = f"/api/pedigrees/{pedigree_id}/import/xeg"
        self._send("POST", path, {"content": xeg_xml})

    def delete_pedigree(self, pedigree_id: str) -> None:
        self._send("DELETE", f"/api/pedigrees/{pedigree_id}", None)

    def _request_json(
        self, method: HttpMethod, path: str, body: dict[str, Any]
    ) -> dict[str, Any]:
        response = self._send(method, path, body)
        try:
            payload = response.json()
        except ValueError as exc:
            raise EvageneApiError(
                f"Evagene API returned non-JSON body for {method} {path}: {exc}"
            ) from exc
        if not isinstance(payload, dict):
            kind = type(payload).__name__
            raise EvageneApiError(
                f"Evagene API returned non-object JSON for {method} {path}: {kind}"
            )
        return payload

    def _send(
        self, method: HttpMethod, path: str, body: dict[str, Any] | None
    ) -> Any:
        url = f"{self._base_url}{path}"
        response = self._http.send(method, url, headers=self._headers(), body=body)
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise EvageneApiError(
                f"Evagene API returned HTTP {response.status_code} for {method} {path}"
            )
        return response

    def _headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }


def _require_str(payload: dict[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str):
        raise EvageneApiError(f"Evagene response is missing string field {key!r}")
    return value

File: family-history-intake-form/node/src/evageneClient.ts
/**
 * Thin client for the subset of the Evagene REST API the intake form needs.
 *
 * One method per endpoint. No orchestration (that lives in IntakeService).
 * All state goes through the injected HttpGateway so tests can swap in a fake.
 */

import type { HttpGateway } from './httpGateway.js';

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class EvageneApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EvageneApiError';
  }
}

export type BiologicalSex = 'female' | 'male' | 'unknown';

export type RelativeType =
  | 'mother'
  | 'father'
  | 'maternal_grandmother'
  | 'maternal_grandfather'
  | 'paternal_grandmother'
  | 'paternal_grandfather'
  | 'sister'
  | 'brother'
  | 'half_sister'
  | 'half_brother';

export interface EvageneClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly http: HttpGateway;
}

export interface CreatePedigreeArgs {
  readonly displayName: string;
}

export interface CreateIndividualArgs {
  readonly displayName: string;
  readonly biologicalSex: BiologicalSex;
  readonly yearOfBirth?: number;
}

export interface AddRelativeArgs {
  readonly pedigreeId: string;
  readonly relativeOf: string;
  readonly relativeType: RelativeType;
  readonly displayName: string;
  readonly biologicalSex: BiologicalSex;
  readonly yearOfBirth?: number;
}

/**
 * The subset of the Evagene API that the intake flow depends on.
 * IntakeService talks to this, not the concrete class, so tests
 * supply their own implementation without a live HTTP layer.
 */
export interface EvageneApi {
  createPedigree(args: CreatePedigreeArgs): Promise<string>;
  createIndividual(args: CreateIndividualArgs): Promise<string>;
  addIndividualToPedigree(pedigreeId: string, individualId: string): Promise<void>;
  designateAsProband(individualId: string): Promise<void>;
  addRelative(args: AddRelativeArgs): Promise<string>;
}

export class EvageneClient implements EvageneApi {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly http: HttpGateway;

  constructor(options: EvageneClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.http = options.http;
  }

  async createPedigree(args: CreatePedigreeArgs): Promise<string> {
    const payload = await this.postJson('/api/pedigrees', { display_name: args.displayName });
    return requireStringField(payload, 'id');
  }

  async createIndividual(args: CreateIndividualArgs): Promise<string> {
    const body: Record<string, unknown> = {
      display_name: args.displayName,
      biological_sex: args.biologicalSex,
    };
    withYearOfBirth(body, args.yearOfBirth);
    const payload = await this.postJson('/api/individuals', body);
    return requireStringField(payload, 'id');
  }

  async addIndividualToPedigree(pedigreeId: string, individualId: string): Promise<void> {
    await this.sendIgnoringBody(
      'POST',
      `/api/pedigrees/${pedigreeId}/individuals/${individualId}`,
      {},
    );
  }

  async designateAsProband(individualId: string): Promise<void> {
    await this.sendIgnoringBody('PATCH', `/api/individuals/${individualId}`, { proband: 1 });
  }

  async addRelative(args: AddRelativeArgs): Promise<string> {
    const body: Record<string, unknown> = {
      relative_of: args.relativeOf,
      relative_type: args.relativeType,
      display_name: args.displayName,
      biological_sex: args.biologicalSex,
    };
    withYearOfBirth(body, args.yearOfBirth);
    const payload = await this.postJson(
      `/api/pedigrees/${args.pedigreeId}/register/add-relative`,
      body,
    );
    const individual = requireObjectField(payload, 'individual');
    return requireStringField(individual, 'id');
  }

  private async postJson(path: string, body: unknown): Promise<Record<string, unknown>> {
    const response = await this.send('POST', path, body);
    return parseObjectBody(response, path);
  }

  private async sendIgnoringBody(
    method: 'POST' | 'PATCH',
    path: string,
    body: unknown,
  ): Promise<void> {
    await this.send(method, path, body);
  }

  private async send(method: 'POST' | 'PATCH', path: string, body: unknown) {
    const url = `${this.baseUrl}${path}`;
    const response = await this.http.send({
      method,
      url,
      headers: this.headers(),
      body,
    });
    if (response.status < HTTP_OK_LOWER || response.status >= HTTP_OK_UPPER) {
      throw new EvageneApiError(
        `Evagene API returned HTTP ${response.status.toString()} for ${method} ${path}`,
      );
    }
    return response;
  }

  private headers(): Record<string, string> {
    return {
      'X-API-Key': this.apiKey,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };
  }
}

async function parseObjectBody(
  response: { json(): Promise<unknown> },
  path: string,
): Promise<Record<string, unknown>> {
  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new EvageneApiError(`Evagene API returned non-JSON body for ${path}: ${reason}`);
  }
  if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
    throw new EvageneApiError(`Evagene API returned non-object JSON for ${path}`);
  }
  return payload as Record<string, unknown>;
}

function withYearOfBirth(body: Record<string, unknown>, year: number | undefined): void {
  if (year === undefined) {
    return;
  }
  body.properties = { year_of_birth: year };
}

function requireStringField(payload: Record<string, unknown>, key: string): string {
  const value = payload[key];
  if (typeof value !== 'string') {
    throw new EvageneApiError(`Evagene response is missing string field '${key}'`);
  }
  return value;
}

function requireObjectField(
  payload: Record<string, unknown>,
  key: string,
): Record<string, unknown> {
  const value = payload[key];
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new EvageneApiError(`Evagene response is missing object field '${key}'`);
  }
  return value as Record<string, unknown>;
}
